stackhpc_ipa_hardware_managers/system_nic.py

```diff
@@ -0,0 +1,241 @@
+# Copyright 2015 Rackspace, Inc.
+# Copyright 2018 StackHPC Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import glob
+import os
+
+from ironic_python_agent import errors
+from ironic_python_agent import hardware
+from ironic_python_agent import utils
+
+from oslo_concurrency import processutils
+from oslo_log import log
+
+UEVENT_PCI_SLOT_NAME_PREFIX = "PCI_SLOT_NAME="
+
+LOG = log.getLogger()
+
+
+def _get_lspci_output(vendor_id=None, device_id=None, **kwargs):
+    device_filter_arg = ''
+    if vendor_id or device_id:
+        device_filter_arg = "-d " + _get_device_filter_string(vendor_id,
+                                                              device_id)
+    lspci_output = ''
+    kwargs["shell"] = True
+    try:
+        lspci_output, _e = utils.execute(
+            "lspci -vmmnD {filter}".format(filter=device_filter_arg), **kwargs)
+    except (processutils.ProcessExecutionError, OSError):
+        raise errors.CleaningError("Unable to run lspci")
+    return lspci_output
+
+
+def _parse_lspci_output(lspci_output):
+    all_records = lspci_output.split("\n\n")
+    devices = []
+    for record_str in all_records:
+        if record_str == "":
+            continue
+        fields_raw = record_str.splitlines()
+        fields = map((lambda x: x.split(':\t')), fields_raw)
+        devices.append(dict(list(map(tuple, fields))))
+    return devices
+
+
+def _get_device_filter_string(vendor_id, device_id):
+    device_id_str = ""
+    device_id_str += vendor_id or ""
+    device_id_str += ":"
+    device_id_str += device_id or ""
+    return device_id_str
+
+
+def _get_base_in_relative_path(path):
+    # given relative path, returns the base element, e.g one/two/three,
+    # returns one
+    if os.path.isabs(path):
+        raise ValueError("path must be relative")
+    dirname = os.path.dirname(path)
+    if dirname == "":
```

Comment:
Maybe I missed something but could you just do something like return path.split('/')[0]?

Reply:
I was trying to keep path handling generic, but maybe using `split` with `os.sep` would be ok.